### ISSUE TITLE
DefinedSchema: Indexes not created correctly for Pointer types

### ISSUE THREAD

**Issue by sadortun** [parse-community/parse-server#8455]
### New Issue Checklist

/cc @Moumouls 

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [x] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases). 

### Issue Description

When creating indexes on `Pointer` fields, the index is not created on the correct field name in `MongoDB`

`Parse` will add a `_p_` prefix to the field name, but `DefinedSchema` won't. This have for consequence to leave the `Pointer` field without an index, and creating an index on a non-existant field.

Had a chat with @Moumouls about this, here is his comment:
> The correct version indeed is the version with the `_p_`, it is a quick fix that was pushed during the pr of the defined schemas. I remember it needed some refactor to allow just the field name without the _p_ because of validators. 

### Steps to reproduce

For example:

```ts
export default SchemaMigrations.makeSchema('Message', {
   fields: {
      recipient: { type: 'Pointer', targetClass: '_User', required: true },
   },
   indexes: {
       recipient: { recipient: 1 },
   },
```

Looking at the indexes:

```
db.getCollection('Message').stats()
{
   "_id_": new NumberInt("749568"),
    "recipient": new NumberInt("90112"),
}
```

In this case, `Parse` will create a DB field named `_p_recipient`, **not** `recipient`.

This have for effect of leaving the  `_p_recipient`  un-indexed, and creating an unused index `recipient`

### Actual Outcome

Index is created on unused field `recipient`

### Expected Outcome

Index need to be created on `_p_recipient`

### Environment

Server
- Parse Server version: `5`
- Operating system: `w11`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `gcloud`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `4.4`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `gcloud`

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): `javascript`
- SDK version: `4`

**Comment by rgunindi** [comment]
I can work on this, but in order to advance, I need additional details about the problem. Any advice or suggestions for me?

**Comment by rgunindi** [comment]
``` it('should create index with _p_ prefix for pointer',async()=>{
      const server = await reconfigureServer();
      const indexes = { complex: {  createdAt: 1, updatedAt: 1, recipient:1 } };
      const schemas = {
        definitions: [{ className: 'Test', fields: { recipient: { type: 'Pointer', targetClass:'_User',require:true } }, indexes }],
      };
      await new DefinedSchemas(schemas, server.config).execute();
      const schema = await new Parse.Schema('Test').get();
      cleanUpIndexes(schema);
      expect(schema.indexes).withContext("Schema indexes did not match expected indexes");
    }) 
```

I ran this test case for this state but couldn't find a bug. Are you certain about this?

**.      ✓ should create index with _p_ prefix for pointer**


@sadortun

**Comment by sadortun** [comment]
@rgunindi 

Can you check : 

1.  `schema.indexes` contains `_p_recipient`



2.  MongoDB console

```
db.getCollection('Test').stats()
{
   "_id_": new NumberInt("1234"),
    "_p_recipient": new NumberInt("1234"),
}
```

**Comment by rgunindi** [comment]
@sadortun 

![image](https://user-images.githubusercontent.com/76232388/224477535-a5f42d73-c11d-4297-bf2a-8cb80c927ea6.png)

```const schemas = {
        definitions: [{
          className: 'Test', fields: {
            recipient: { type: 'Pointer', targetClass: '_User', required: true },
          },
          indexes: {
            recipient: { recipient: 1 },
          }
        }
        ],
      };
```

I re-run and the result same 

![image](https://user-images.githubusercontent.com/76232388/224477830-a58e4cce-a16e-4cec-819a-21b908e48d51.png)

where is the `_p_recipient ?`

**Comment by sadortun** [comment]
@rgunindi 

If you add data, you'll see in MongoDB the pointer field will be in the  `_p_recipient` field. But the index is created on `recipient`.

So, the actual pointer will not be indexed.

**Comment by Moumouls** [comment]
Hi @sadortun @rgunindi , @sadortun is right using a "classic" index on Pointers in a Parse Schema will not work.

I added a quick fix when i worked on Defined Schemas, it's documented in the index section of the Defined Schema docs see: https://docs.parseplatform.org/defined-schema/guide/#indexes

I don't remember exactly but it was a quick fix and the PR to correctly support indexes on pointers without "_p_" was a little bit heavier than supporting the "_p_"

But feel free to open a PR to correctly support the pointer index, i'll be happy to review it !

**Comment by rgunindi** [comment]
@Moumouls @sadortun 
![resim](https://github.com/parse-community/parse-server/assets/76232388/769b3530-d966-4e84-9f57-bfcc6ceb0c38)

That's the way it should be, isn't it?

**Comment by Moumouls** [comment]
Yes @rgunindi if you have a Parse Pointer on the recipient field !

**Comment by rgunindi** [comment]
@Moumouls 
I have applied the fix for this issue and opened pr. Thank you for your patience. Please report deficiencies

[Pr Link](https://github.com/parse-community/parse-server/pull/8533)

**Comment by Moumouls** [comment]
@mtrezza  i think this could be closed, a fix is documented about using indexes on pointers, it's not the best DX, but it works fine and is documented: https://docs.parseplatform.org/defined-schema/guide/#indexes

**Comment by mtrezza** [comment]
What about the [PR](https://github.com/parse-community/parse-server/pull/8533) that has been submitted?

**Comment by mtrezza** [comment]
@coderabbitai 

The docs say:

> To define an index on a Pointer field you need to use the special notation _p_<FIELDNAME>. For example, if you define city: { type: "Pointer", targetClass: "City" } in your fields you can define an index on this pointer with cityIndexExample: { _p_city: true }.

This sounds strange. A developer shouldn't have to care about Parse internals when using a public API. Where is the best place to fix this? Check whether the Parse JS SDK already has an API to create an index and what syntax is used there. Should the fix be implemented in the "defined schema" feature or on the DB controller level, or somewhere else?

**Comment by mtrezza** [comment]
@coderabbitai this sounds like a breaking change. What about the indices for the auth providers which are stored in a special way, how would a developer create such an index?

**Comment by Moumouls** [comment]
@mtrezza  may be we need a new PR to see the end result, with "transformIndexKeys" also defined schema use a "get" request for indexes the transformation need to be 2 way

**Comment by mtrezza** [comment]
I closed https://github.com/parse-community/parse-server/pull/8533; we'll keep this issue open.

**Comment by Moumouls** [comment]
okay thanks @mtrezza , the problem here is not straight forward to fix and need special care
